stdlib/src/base64/_b64encode.mojo

```diff
@@ -23,209 +24,95 @@ Instructions, ACM Transactions on the Web 12 (3), 2018.
+    fn combine[
+        mask: Bytes[4], shift: Int
+    ](shuffled: Bytes[width]) -> Bytes[width]:
+        var `6bit` = shuffled & _repeat_until[width](mask)
+        return _rshift_bits_in_u16[shift](`6bit`)
+
+    var shuffled = input.shuffle[mask = indices()]()
+    var a = combine[
+        Bytes[4](0b0000_0000, 0b1111_1100, 0b0000_0000, 0b0000_0000), 10
+    ](shuffled)
+    var b = combine[
+        Bytes[4](0b1111_0000, 0b0000_0011, 0b0000_0000, 0b0000_0000), -4
+    ](shuffled)
+    var c = combine[
+        Bytes[4](0b0000_0000, 0b0000_0000, 0b1100_0000, 0b0000_1111), 6
+    ](shuffled)
+    var d = combine[
+        Bytes[4](0b0000_0000, 0b0000_0000, 0b0011_1111, 0b0000_0000), 8
+    ](shuffled)
+    return a | b | c | d
+
+
+# | 6-bit Value | ASCII Range | Target index | Offset (6-bit to ASCII) |
+# |-------------|-------------|--------------|-------------------------|
+# |  0 ... 25   | A ... Z     | 13           | 65                      |
+# | 26 ... 51   | a ... z     |  0           | 71                      |
+# | 52 ... 61   | 0 ... 9     |  1 ... 10    | -4                      |
+# | 62          | +           | 11           | -19                     |
+# | 63          | /           | 12           | -16                     |
 # fmt: off
-alias TABLE_BASE64_OFFSETS = SIMD[DType.uint8, 16](
+alias UNUSED = 0
+alias OFFSETS = Bytes[16](
     71,                                     # a ... z
-    -4, -4, -4, -4, -4, -4, -4, -4, -4, -4, # 0 ... 9 
+    -4, -4, -4, -4, -4, -4, -4, -4, -4, -4, # 0 ... 9
     -19,                                    # +
     -16,                                    # /
     65,                                     # A ... Z
     UNUSED, UNUSED
 )
-# fmt: on
 alias END_FIRST_RANGE = 25
 alias END_SECOND_RANGE = 51
+# fmt: on
 
 
-fn _base64_simd_mask[
-    simd_width: Int
-](nb_value_to_load: Int) -> SIMD[DType.bool, simd_width]:
-    alias mask = _compile_time_iota[DType.uint8, simd_width]()
-    return mask < UInt8(nb_value_to_load)
-
-
-fn _repeat_until[
-    dtype: DType, input_size: Int, //, target_size: Int
-](vector: SIMD[dtype, input_size]) -> SIMD[dtype, target_size]:
-    @parameter
-    if target_size == input_size:
-        var same_vector = rebind[SIMD[dtype, target_size]](vector)
-        return same_vector
-    else:
-        return _repeat_until[target_size](vector.join(vector))
-
-
-fn _move_first_group_of_6_bits[
-    simd_width: Int
-](shuffled_vector: SIMD[DType.uint8, simd_width]) -> SIMD[
-    DType.uint8, simd_width
-]:
-    alias mask_1 = _repeat_until[simd_width](
-        SIMD[DType.uint8, 4](0b11111100, 0, 0, 0)
-    )
-    var masked_1 = shuffled_vector & mask_1
-    var result = masked_1 >> 2
-    return result
-
-
-fn _move_second_group_of_6_bits[
-    simd_width: Int
-](shuffled_vector: SIMD[DType.uint8, simd_width]) -> SIMD[
-    DType.uint8, simd_width
-]:
-    alias mask_2 = _repeat_until[simd_width](
-        SIMD[DType.uint8, 4](0b00000011, 0b11110000, 0, 0)
```

Comment:
Can you detail what you mean? When processing a stream of independent bytes, endianness doesn't matter. Endianness only becomes relevant when we’re interpreting multiple bytes as a single larger data type, such as a 16-bit, 32-bit, or 64-bit integer. So unless I'm missing something, things should work out, even with the bitcast afterwards.

Reply:
> Endianness only becomes relevant when we’re interpreting multiple bytes as a single larger data type.

Indeed, we bit-cast to `u16` for the bit rotation since some operations cross byte boundaries. I initially thought your implementation worked because endianness effects canceled out on both ends, yielding a functional result. Now I see I was mistaken.

The current setup does have a few subtle advantages. We start with:
```
|b₃b₂b₁b₀. . . . |a₅a₄a₃a₂a₁a₀b₅b₄|
```
and aim for:
```
|. . b₅b₄b₃b₂b₁b₀|. . a₅a₄a₃a₂a₁a₀|
```
In your original approach, they are shuffled into:
```
|. . . . . . . . |a₅a₄a₃a₂a₁a₀. . |  # rotate right by 2
|b₃b₂b₁b₀. . . . |. . . . . . b₅b₄|  # rotate right by 4
```
A 16-bit rotation positions them correctly, but since these bits aren’t consecutive, the compiler fails to find certain optimisations (like `vpmulhuw` Lemire2018 or `vpmultishiftqb` in Lemire2020). In the paper and the current setup, `a` and `b` are masked so that their bits are consecutive:
```
|a₅a₄a₃a₂a₁a₀. . |. . . . . . . . |  # rotate right by 10
|. . . . . . b₅b₄|b₃b₂b₁b₀. . . . |  # rotate left by 4
```

In any case, thank you for your work on this and for laying such a solid foundation for further optimizations!

Reply:
Thanks, it's possible I made a mistake there. If I understand what you are saying, the code I wrote previously shouldn't have worked in a little-endian system. But the unit test are passing and the CI is little endian. So I guess the code was still correct, no?

Reply:
Sorry, I wasn’t very clear in my wording. Your code was indeed **correct all along**! The issue was my misunderstanding of why it worked (when I first ask the question [here](https://github.com/modularml/mojo/pull/3737#discussion_r1828090761)). I thought I was fixing a subtle bug, but in reality, there was nothing to fix.